File: neet_code_150_cpp/appendix/closed_paren.cpp
#include <iostream>
#include <string>
#include <stack>

bool is_valid_paren(std::string& s) {
    std::stack<char> stack;
    for(auto&c: s) {
        if(!stack.empty()) {
            auto stack_top = stack.top();
            if(stack_top == '(' && c == ')') {
                stack.pop();
            } else {
                stack.push(c);
            }
        } else {
            stack.push(c);
        }
    }

    return stack.empty();
}

int main(void) {
    std::string case_1 = "(())";
    // => true
    std::string case_2 = "()()()";
    // => true
    std::string case_3 = ")(";
    // => false
    std::string case_4 = "(()()";
    // => false

    std::cout << is_valid_paren(case_1) << std::endl;
    std::cout << is_valid_paren(case_2) << std::endl;
    std::cout << is_valid_paren(case_3) << std::endl;
    std::cout << is_valid_paren(case_4) << std::endl;
}

File: neet_code_150/src/appendix/closed_paren.rs
// カッコ'()'が閉じているか否か判定する
// stackの問題の典型例
pub fn valid_paren(p_str: String) -> bool {
    let mut stack = vec![];
    for c in p_str.chars() {
        if let Some(stack_top) = stack.last() {
            match (stack_top, c) {
                ('(', ')') => {
                    stack.pop();
                }
                _ => {
                    stack.push(c);
                }
            }
        } else {
            stack.push(c);
        }
    }

    stack.is_empty()
}

fn main() {
    let case_1 = "(())".to_string();
    // => true
    let case_2 = "()()()".to_string();
    // => true
    let case_3 = ")(".to_string();
    // => false
    let case_4 = "(()()".to_string();
    // => false

    println!("case_1: {}", valid_paren(case_1.clone()));
    println!("case_2: {}", valid_paren(case_2.clone()));
    println!("case_3: {}", valid_paren(case_3.clone()));
    println!("case_4: {}", valid_paren(case_4.clone()));
}
